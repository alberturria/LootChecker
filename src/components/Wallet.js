import React, { Component } from 'react';
import { connect } from 'react-redux';
import { deposit, withdraw } from '../actions/balance';

export class Wallet extends Component {
    constructor() {
        super();

        this.state = { balance: undefined };
    }

    _updateBalance = (event) => {
        this.setState({ balance: parseInt(event.target.value, 10) });
    }

    _deposit = () => this.props.deposit(this.state.balance);

    _withdraw = () => this.props.withdraw(this.state.balance);

    render() {
        return(
            <div>
                <h3 className='balance'>Wallet balance: {this.props.balance}</h3>
                <br />
                <input className='input-wallet' onChange={this._updateBalance}/>
                <button className='btn-deposit' onClick={this._deposit}>Deposit</button>
                <button className='btn-withdraw' onClick={this._withdraw}>Withdraw</button>
            </div>
        );
    }
}

export default connect(state => { return { balance: state } }, { deposit, withdraw })(Wallet);