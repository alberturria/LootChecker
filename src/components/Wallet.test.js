import React from 'react';
import { shallow } from 'enzyme';
import { Wallet } from './Wallet';

describe('Wallet', () => {
    const mockDeposit = jest.fn();
    const mockWithdraw = jest.fn();
    const props = { balance: 20, deposit: mockDeposit, withdraw: mockWithdraw };
    const wallet = shallow(<Wallet {...props}/>);
    it('renders properly', () => {
        expect(wallet).toMatchSnapshot();
    });

    it('displays the balance from props', () => {
        const expectedBalance = 'Wallet balance: 20';
        const balanceNode = wallet.find('.balance');
        expect(balanceNode.text()).toEqual(expectedBalance);
    });

    it('creates an input to deposit into or withdraw from the balance', () => {
        const inputNode = wallet.find('.input-wallet');
        expect(inputNode.exists()).toBe(true);
    })

    describe('When the user types into the wallet input', () => {
        const userBalance = '25';
        
        beforeEach(() => {
            const inputNode = wallet.find('.input-wallet');
            inputNode.simulate('change', { target: { value: userBalance }});
        });

        it('updates the local wallet balance in `state` and converts it to a number', () => {
            expect(wallet.state().balance).toEqual(parseInt(userBalance, 10));
        })

        describe('and the user wants to make a deposit', () => {
            beforeEach(() => {
                const depositButton = wallet.find('.btn-deposit')
                depositButton.simulate('click');
            })
            
            it('dispatches the `deposit()` it receives from props with local balance', () => {
                expect(mockDeposit).toHaveBeenCalledWith(parseInt(userBalance, 10));
            });
        })

        describe('and the user wants to make a withdrawl', () => {
            beforeEach(() => {
                const withdrawButton = wallet.find('.btn-withdraw')
                withdrawButton.simulate('click');
            })

            it('dispatches the `withdraw()` method it receives from props with the local balance', () => {
                expect(mockWithdraw).toHaveBeenCalledWith(parseInt(userBalance, 10));
            })
        })
    });
})