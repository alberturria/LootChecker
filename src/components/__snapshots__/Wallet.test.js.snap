// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Wallet renders properly 1`] = `
<div>
  <h3
    className="balance"
  >
    Wallet balance: 
    20
  </h3>
  <br />
  <input
    className="input-wallet"
    onChange={[Function]}
  />
  <button
    className="btn-deposit"
    onClick={[Function]}
  >
    Deposit
  </button>
  <button
    className="btn-withdraw"
    onClick={[Function]}
  >
    Withdraw
  </button>
</div>
`;
